pex/pex_bootstrapper.py

```diff
@@ -56,28 +61,100 @@ def find_in_path(target_interpreter):
       return try_path
 
 
-def maybe_reexec_pex():
-  from .variables import ENV
-  if not ENV.PEX_PYTHON:
-    return
+def _get_python_interpreter(binary):
+  try:
+    return PythonInterpreter.from_binary(binary)
+  # from_binary attempts to execute the interpreter
+  except Executor.ExecutionError:
+    return None
+
+
+def _find_compatible_interpreters(pex_python_path, compatibility_constraints):
+  """Find all compatible interpreters on the system within the supplied constraints and use
+     PEX_PYTHON_PATH env variable if it is set. If not, fall back to interpreters on $PATH.
+  """
+  if pex_python_path:
+    interpreters = []
+    for binary in pex_python_path.split(os.pathsep):
+      pi = _get_python_interpreter(binary)
+      if pi:
+        interpreters.append(pi)
+    if not interpreters:
+      die('No interpreters from PEX_PYTHON_PATH can be found on the system. Exiting.')
+  else:
+    # All qualifying interpreters found in $PATH
+    interpreters = PythonInterpreter.all()
 
-  from .common import die
-  from .tracer import TRACER
+  compatible_interpreters = list(matched_interpreters(
+    interpreters, compatibility_constraints, meet_all_constraints=True))
+  return compatible_interpreters if compatible_interpreters else None
 
-  target_python = ENV.PEX_PYTHON
+
+def _handle_pex_python(target_python, compatibility_constraints):
+  """Re-exec using the PEX_PYTHON interpreter"""
   target = find_in_path(target_python)
   if not target:
     die('Failed to find interpreter specified by PEX_PYTHON: %s' % target)
+  elif compatibility_constraints:
+    pi = PythonInterpreter.from_binary(target)
+    if not all(pi.identity.matches(constraint) for constraint in compatibility_constraints):
+      die('Interpreter specified by PEX_PYTHON (%s) is not compatible with specified '
+          'interpreter constraints: %s' % (target, str(compatibility_constraints)))
   if os.path.exists(target) and os.path.realpath(target) != os.path.realpath(sys.executable):
     TRACER.log('Detected PEX_PYTHON, re-exec to %s' % target)
-    ENV.delete('PEX_PYTHON')
+    ENV.SHOULD_EXIT_BOOTSTRAP_REEXEC = True
     os.execve(target, [target_python] + sys.argv, ENV.copy())
 
 
+def _handle_general_interpreter_selection(pex_python_path, compatibility_constraints):
+  """Handle selection in the case that PEX_PYTHON_PATH is set or interpreter compatibility
+     constraints are specified.
+  """
+  compatible_interpreters = _find_compatible_interpreters(
+    pex_python_path, compatibility_constraints)
+
+  target = None
+  if compatible_interpreters:
+    target = min(compatible_interpreters).binary
+  if not target:
+    die('Failed to find compatible interpreter for constraints: %s'
+        % str(compatibility_constraints))
```

Comment:
seems like this `die` should be raised `if not compatible_interpreters` vs `if not target`. e.g.

```
compatible_interpreters = ...
if not compatible_interpreters:
  die(...)
target = min(compatible_interpreters).binary
```

unless its possible for `min(...).binary` to also be `''` or `None`?